File: src/python/benchmark/hdf5_benchmark.py
from __future__ import print_function

import nifty
import nifty.hdf5
import h5py
import threading

fileName = "/home/tbeier/Desktop/ufd_overseg6.h5"
#fileName = "/home/tbeier/_4nm_ufd_overseg.h5"
dsetName = "data"

shape = [2000,2000,2000]
blockShape = [100,100,100]



blocking = nifty.tools.blocking(roiBegin=[0]*3, roiEnd=shape, 
                                blockShape=blockShape)

numberOfBlocks = blocking.numberOfBlocks

nThreads = 40



if True:
    # now the same in python
    h5File = h5py.File(fileName,'r')
    array = h5File[dsetName]
    lock = threading.Lock()

    with nifty.Timer("python"):
        val = long(0)
        def f(blockIndex):
            global val
            block = blocking.getBlock(blockIndex)
            b,e = block.begin, block.end

            with lock:
                subarray = array[b[0]:e[0], b[1]:e[1], b[2]:e[2]]
                _val = subarray[0,0,0]
                val += long(_val)

        nifty.tools.parallelForEach(range(numberOfBlocks),f=f, nWorkers=nThreads)
        print("val",val)

print()

if True:
    # first c++
    h5File = nifty.hdf5.openFile(fileName)
    array = nifty.hdf5.Hdf5ArrayUInt32(h5File, dsetName)

    with nifty.Timer("c++"):
        nifty.hdf5.runBenchmark(array, blocking, nThreads)

print()

if True:
    # first c++
    h5File = nifty.hdf5.openFile(fileName)
    array = nifty.hdf5.Hdf5ArrayUInt32(h5File, dsetName)
    lock = threading.Lock()
    with nifty.Timer("c++ python"):
        val = long(0)
        def f(blockIndex):
            global val
            block = blocking.getBlock(blockIndex)
            b,e = block.begin, block.end

            with lock:
                subarray = array[b[0]:e[0], b[1]:e[1], b[2]:e[2]]
                _val = subarray[0,0,0]
                val += long(_val)

        nifty.tools.parallelForEach(range(numberOfBlocks),f=f, nWorkers=nThreads)
        print("val",val)

File: src/python/lib/hdf5/hdf5_benchmark.cxx
#ifdef WITH_HDF5

#include <pybind11/pybind11.h>
#include <iostream>
#include <sstream>
#include <pybind11/numpy.h>
#include <pybind11/stl.h>

#include "nifty/python/converter.hxx"
#include "nifty/hdf5/hdf5_array.hxx"
#include "nifty/tools/blocking.hxx"
#include "nifty/parallel/threadpool.hxx"

namespace py = pybind11;



namespace nifty{
namespace hdf5{


    void exportBenchmark(py::module & hdf5Module) {


        hdf5Module.def("runBenchmark",
        [](
            const nifty::hdf5::Hdf5Array<uint32_t> & data,
            const nifty::tools::Blocking<3> & blocking,
            const int numberOfThreads
        ){

            std::mutex lock;
            uint64_t val = 0;

            const auto numberOfBlocks = blocking.numberOfBlocks();
            nifty::parallel::parallel_foreach(
                numberOfThreads,
                numberOfBlocks,
                [&](
                    const int tid,
                    const int blockIndex
                ){
                    typedef typename xt::xarray<uint32_t>::shape_type ShapeType;
                    const auto block = blocking.getBlock(blockIndex);
                    const ShapeType blockShape(block.shape().begin(), block.shape().end());
                    lock.lock();
                    xt::xarray<uint32_t> subarray(blockShape);
                    data.readSubarray(block.begin().begin(), subarray);

                    const auto _val = subarray(0,0,0);

                    val += _val;
                    lock.unlock();
                }
            );
            // to make sure the above code is not optimized away
            std::cout<<"val "<<val<<"\n";
        }
        );

    }

}
}

#endif
